### PR TITLE
StreamInfo: Add downstreamDirectRemoteAddress

### LINKED ISSUE #4996: StreamInfo: Preserving the source address on downstreamRemoteAddress
*Description*:
Our project's logging requirements call for the direct connected host, i.e. the actual physical address on the other side of the downstream connection. The logging interface includes only request headers, response headers/trailers, and StreamInfo. StreamInfo.downstreamRemoteAddress() contains this information at first, but it's potentially modified by x-forwarded-for or a proxy proto.

A couple of proposals:

- Storing the downstream Network::Connection object in StreamInfo, which would be useful if anyone in the future needs information from the connection. Connection info could also be loggable via a change to the logging interface.
- Making the downstreamRemoteAddress behaviour configurable, i.e. a flag that will prevent it from being overwritten by proxy or forwarded-for info
- Adding another field to StreamInfo, e.g. downstreamDirectlyConnectedAddress

Thoughts? Opinions? Points of view?

### PR DESCRIPTION
**Pull request by auni53**
Signed-off-by: Auni Ahsan <auni@google.com>

*Description*: The downstreamRemoteAddress field is modified when the remote address is inferred from proxy information or headers. This new field guarantees to hold the directly connected host's physical address.
*Risk Level*: Low
*Testing*: Searched around for references to downstreamRemoteAddress() and filled in similar tests/mocks.

Fixes #4996.

### CODE REVIEW

**Comment by mattklein123** [comment]
@dnoe PTAL

**Comment by auni53** [comment]
what is PTAL?

**Comment by dnoe** [comment]
I'm going to have to punt this off on someone else since I have a critical issue on my plate this week. Thanks for taking an initial look @venilnoronha - let me know when it is ready from your standpoint and I will move it forward.

**Comment by venilnoronha** [comment]
@dnoe will do.

**Comment by auni53** [comment]
Sorry for the delay, was out sick last week. There's no test ascertaining the changing state for downstreamRemoteAddress(), so I tacked on w xff to a conn manager impl test.

**Comment by auni53** [comment]
Sorry, this branch is getting a little messy. Forgot to signoff like 100 commits ago before merging, gimme a few to clean this up. (I miss rebasing.)

**Comment by htuch** [comment]
@dnoe no worries, I can review.

**Comment by htuch** [comment]
@auni53 there are build issues here under ASAN and other CI tests, can you take a look?

File: source/common/http/conn_manager_impl.cc [hunk 1]
```diff
@@ -369,6 +369,8 @@ ConnectionManagerImpl::ActiveStream::ActiveStream(ConnectionManagerImpl& connect
   }
   stream_info_.setDownstreamLocalAddress(
       connection_manager_.read_callbacks_->connection().localAddress());
+  stream_info_.setDownstreamDirectlyConnectedAddress(
```
Comment (htuch) [inline source/common/http/conn_manager_impl.cc]: I think I would prefer `downstreamDirectRemoteAddress()` for consistency with existing naming, or something like that.

File: test/common/http/conn_manager_impl_test.cc [hunk 2]
```diff
@@ -932,6 +932,11 @@ TEST_F(HttpConnectionManagerImplTest,
 }
 
 TEST_F(HttpConnectionManagerImplTest, TestAccessLog) {
+  constexpr char local_address[] = "0.0.0.0";
+  constexpr char xff_address[] = "1.2.3.4";
+
+  // stream_info.directRemoteAddress() will populate from xff
```
Comment (htuch) [inline test/common/http/conn_manager_impl_test.cc]: `directRemoteAddress`? How does this populate from XFF?

File: test/common/http/conn_manager_impl_test.cc [hunk 3]
```diff
@@ -932,6 +932,11 @@ TEST_F(HttpConnectionManagerImplTest,
 }
 
 TEST_F(HttpConnectionManagerImplTest, TestAccessLog) {
+  constexpr char local_address[] = "0.0.0.0";
```
Comment (htuch) [inline test/common/http/conn_manager_impl_test.cc]: Ideally we also have a more specific unit test on `ConnectionManagerImpl::ActiveStream::ActiveStream`, i.e. HCM, that shows this behavior. If there's a reasonable way to add this without too many gymnastics, that could be great.

**Comment by auni53** [comment]
Kinda bizarre...the build failures are actually just clang-tidy failures, but they only throw on asan/tsan/mac runs. (Still wish there was an easier way to validate these errors than waiting a long time.)

File: docs/root/intro/version_history.rst [hunk 4]
```diff
@@ -68,6 +68,7 @@ Version history
 * stream: renamed the `RequestInfo` namespace to `StreamInfo` to better match
   its behaviour within TCP and HTTP implementations.
 * stream: renamed `perRequestState` to `filterState` in `StreamInfo`.
+* stream: added field `downstreamDirectRemoteAddress` and setter to `StreamInfo`.
```
Comment (venilnoronha) [inline docs/root/intro/version_history.rst]: nit: "and setter" can be skipped.

File: include/envoy/stream_info/stream_info.h [hunk 5]
```diff
@@ -276,16 +276,30 @@ class StreamInfo {
    */
   virtual const Network::Address::InstanceConstSharedPtr& downstreamLocalAddress() const PURE;
 
+  /**
+   * @param downstream_directly_connected_address sets the direct physical address of downstream
+   * connection.
+   */
+  virtual void setDownstreamDirectRemoteAddress(
+      const Network::Address::InstanceConstSharedPtr& downstream_directly_connected_address) PURE;
```
Comment (venilnoronha) [inline include/envoy/stream_info/stream_info.h]: The variable name needs to change to something like `downstream_direct_remote_address`.

File: source/common/stream_info/stream_info_impl.h [hunk 6]
```diff
@@ -219,6 +228,7 @@ struct StreamInfoImpl : public StreamInfo {
   uint64_t bytes_sent_{};
   Network::Address::InstanceConstSharedPtr upstream_local_address_;
   Network::Address::InstanceConstSharedPtr downstream_local_address_;
+  Network::Address::InstanceConstSharedPtr downstream_directly_connected_address_;
```
Comment (venilnoronha) [inline source/common/stream_info/stream_info_impl.h]: Here too.

File: test/common/stream_info/test_util.h [hunk 7]
```diff
@@ -190,6 +197,7 @@ class TestStreamInfo : public StreamInfo::StreamInfo {
   bool hc_request_{};
   Network::Address::InstanceConstSharedPtr upstream_local_address_;
   Network::Address::InstanceConstSharedPtr downstream_local_address_;
+  Network::Address::InstanceConstSharedPtr downstream_directly_connected_address_;
```
Comment (venilnoronha) [inline test/common/stream_info/test_util.h]: And here.

File: test/mocks/stream_info/mocks.h [hunk 8]
```diff
@@ -88,6 +92,7 @@ class MockStreamInfo : public StreamInfo {
   uint64_t bytes_sent_{};
   Network::Address::InstanceConstSharedPtr upstream_local_address_;
   Network::Address::InstanceConstSharedPtr downstream_local_address_;
+  Network::Address::InstanceConstSharedPtr downstream_directly_connected_address_;
```
Comment (venilnoronha) [inline test/mocks/stream_info/mocks.h]: And this one.

**Comment by venilnoronha** [comment]
@auni53 you can ignore the clang_tidy error. It's just a timeout.

```
ci/build_setup.sh: line 115:   199 Hangup                  ./run_clang_tidy.sh
Too long with no output (exceeded 10m0s)
```

@htuch over to you for another pass.

**Comment by auni53** [comment]
@lizan could you take a look at this? The clang_tidy tests on circleCI time out, and my local run isn't showing me anything useful.

Local run details: I ran `./ci/run_envoy_docker.sh ci/do_ci.sh bazel.clang_tidy`; it compiled the DB and had a bunch of "x warnings generated" messages without any errors. The beginning of the run included this error/warning repeated a few times:
```$TEST_TMPDIR defined: output root default is '/build/tmp' and max_idle_secs default is '15'.
WARNING: Duplicate rc file: /source/tools/bazel.rc is read multiple times, most recently imported from /source/ci/.bazelrc
WARNING: Processed legacy workspace file /source/ci/tools/bazel.rc. This file will not be processed in the next release of Bazel. Please read https://github.com/bazelbuild/bazel/issues/6319 for further information, including how to upgrade.
```

**Comment by lizan** [comment]
@auni53 can you merge master to see how it goes? I add an echo for clang-tidy to identify the problem.

**Comment by auni53** [comment]
doneeee

**Comment by htuch** [comment]
@lizan still not seeing any useful debug here. Do we just want to force merge this PR or is this a useful reproducer to keep investigating with?

**Comment by lizan** [comment]
@htuch I opened #5296 to debug CI, if you want to merge this PR sooner, then I think it is OK to force merge.

**Comment by htuch** [comment]
@lizan ok, will force merge when a reasonable number of these presubmits pass.

**Pull request merged**